indra/llcommon/lua_function.h

```diff
@@ -226,178 +240,45 @@ std::string lua_emplace_metaname(const std::string& Tname = LLError::Log::classn
-    // retrieve the first (only) bound upvalue and push to stack top
-    lua_pushvalue(L, lua_upvalueindex(1));
-    // This is the weak_table bound by lua_emplace<T>(). Its one and only
-    // entry should be the lua_emplace<T>() userdata -- unless userdata has
-    // been garbage collected. Retrieve weak_table[1].
-    lua_pushinteger(L, 1);
-    // stack contains weak_table, 1
-    lua_gettable(L, -2);
-    // stack contains weak_table, weak_table[1]
-    // If our userdata was garbage-collected, there is no weak_table[1],
-    // and we just retrieved nil.
-    if (lua_isnil(L, -1))
-    {
-        lua_pop(L, 2);
-        return 0;
-    }
-    // stack contains weak_table, userdata
-    // ditch weak_table
-    lua_replace(L, -2);
-    // pass userdata to lua_emplace_gc<T>()
-    return lua_emplace_gc<T>(L);
-}
-
-// set as metatable(userdata).__gc to be called by the garbage collector
-template <class T>
-int lua_emplace_gc(lua_State* L)
-{
-    using optT = std::optional<T>;
-    // We're called with userdata on the stack holding an instance of type T.
-    auto ptr = lua_touserdata(L, -1);
-    llassert(ptr);
-    // Destroy the T object contained in optT at the void* address ptr. If
-    // in future lua_emplace() must manually align our optT* within the
-    // Lua-provided void*, derive optT* from ptr.
-    static_cast<optT*>(ptr)->reset();
-    // pop the userdata
-    lua_pop(L, 1);
-    return 0;
-}
-
-template <class T>
-std::string lua_emplace_metaname(const std::string& Tname)
-{
-    return stringize("lua_emplace_", Tname, "_meta");
-}
-
-} // anonymous namespace
-
 /**
  * If the value at the passed acceptable index is a full userdata created by
- * lua_emplace<T>() -- that is, the userdata contains a non-empty
- * std::optional<T> -- return a pointer to the contained T instance. Otherwise
- * (index is not a full userdata; userdata is not of type std::optional<T>;
- * std::optional<T> is empty) return nullptr.
+ * lua_emplace<T>(), return a pointer to the contained T instance. Otherwise
+ * (index is not a full userdata; userdata is not of type T) return nullptr.
  */
 template <class T>
 T* lua_toclass(lua_State* L, int index)
 {
-    using optT = std::optional<T>;
-    // recreate the name lua_emplace<T>() uses for its metatable
-    auto metaname{ lua_emplace_metaname<T>() };
     // get void* pointer to userdata (if that's what it is)
-    void* ptr{ luaL_checkudata(L, index, metaname.c_str()) };
-    if (! ptr)
-        return nullptr;
-    // Derive the optT* from ptr. If in future lua_emplace() must manually
-    // align our optT* within the Lua-provided void*, adjust accordingly.
-    optT* tptr(static_cast<optT*>(ptr));
-    // make sure our optT isn't empty
-    if (! *tptr)
-        return nullptr;
-    // looks like we still have a non-empty optT: return the *address* of the
-    // value() reference
-    return &tptr->value();
+    void* ptr{ lua_touserdatatagged(L, index, TypeTag<T>::value) };
```

Comment:
Ok, I didn't know about ` lua_touserdatatagged` before. Looks quite convenient.

Reply:
I didn't either! Joosten has been enlightening me about various of the Luau extensions over PUC-Rio Lua.